File: code/649-dota2-senate.go
package main

import "fmt"

func predictPartyVictory(senate string) string {
	cnt := 0
	seenR, seenD := true, true
	ss := []rune(senate)
	for seenR && seenD {
		seenR, seenD = false, false
		for i := 0; i < len(ss); i++ {
			if ss[i] == 'R' {
				if cnt < 0 {
					ss[i] = 'O'
				}
				cnt++
				seenR = true
			} else if ss[i] == 'D' {
				if cnt > 0 {
					ss[i] = 'O'
				}
				cnt--
				seenD = true
			}
		}
	}
	if seenR {
		return "Radiant"
	} else {
		return "Dire"
	}
}

func main() {
	// senate := "RD"
	senate := "RDD"
	fmt.Println(predictPartyVictory(senate))
}

File: code/649-dota2-senate.py
from collections import deque


class Solution:
    def predictPartyVictory(self, senate: str) -> str:
        queue = deque()
        people = [0, 0]
        ban = [0, 0]

        for p in senate:
            x = int(p == 'R')
            people[x] += 1
            queue.append(x)

        while all(people):
            x = queue.popleft()
            if ban[x] > 0:
                ban[x] -= 1
                people[x] -= 1
            else:
                ban[1 - x] += 1
                queue.append(x)
        return "Radiant" if people[1] else "Dire"
